src/eve/traits.py

```diff
@@ -20,73 +20,124 @@
+            return instance.collected
+
+
+class SymbolRefsValidatorTrait(concepts.Model):
+    """Node trait adding automatic validation of symbol references appearing the node tree.
+
+    It assumes that the symbol table with the actual definitions is stored as
+    a dict in the `symtable_` attribute (like :class:`SymbolTableTrait` does).
+    """
+
+    @pydantic.root_validator(skip_on_failure=True)
+    def __validate_refs(  # type: ignore  # validators are classmethods
+        cls: Type[SymbolRefsValidatorTrait], values: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        validator = cls._SymbolRefsValidator()
+        for v in values.values():
+            validator.visit(v, symtable=values["symtable_"])
+
+        if len(validator.missing_symbols) > 0:
+            raise ValueError("Symbols {} not found.".format(validator.missing_symbols))
+
         return values
 
-    def collect_symbols(self) -> None:
-        self.symtable_ = dict()
-        self.symtable_ = self._collect_symbols(self)
+    class _SymbolRefsValidator(visitors.NodeVisitor):
+        def __init__(self) -> None:
+            self.missing_symbols: Set[str] = set()
 
-    @staticmethod
-    @contextlib.contextmanager
-    def symtable_merger(
-        node_visitor: visitors.NodeVisitor, node: concepts.Node, kwargs: Dict[str, Any]
-    ) -> Iterator[None]:
-        """Update or add the symtable to kwargs in the visitor calls.
+        def visit_Node(
+            self, node: concepts.Node, *, symtable: Dict[str, Any], **kwargs: Any
+        ) -> None:
+            for name, metadata in node.__node_children__.items():
+                if isinstance(metadata["definition"].type_, type) and issubclass(
+                    metadata["definition"].type_, SymbolRef
+                ):
+                    if getattr(node, name) and getattr(node, name) not in symtable:
+                        self.missing_symbols.add(getattr(node, name))
 
-        This is a context manager that, when included to the contexts classvar, will
-        automatically pass 'symtable' as a keyword argument to visitor methods.
-        """
+            if isinstance(node, SymbolTableTrait):
+                symtable = {**symtable, **node.symtable_}
+            self.generic_visit(node, symtable=symtable, **kwargs)
+
+
+class ValidatedSymbolTableTrait(SymbolRefsValidatorTrait, SymbolTableTrait):
+    """Node trait adding an automatically created and validated symbol table.
+
+    It is just the combination of the :class:`SymbolTableTrait` and
+    :class:`SymbolRefsValidatorTrait` traits.
+    """
+
+    pass
+
+
+# --- Visitor Traits ---
+class VisitorWithSymbolTableTrait(visitors.NodeVisitor):
+    """Visitor trait to update or add the symtable to kwargs in the visitor calls.
+
+    Visitors inhering from this trait will automatically pass the active
+    symbol table to visitor methods as the 'symtable' keyword argument.
+    """
+
+    def visit(self, node: concepts.TreeNode, **kwargs: Any) -> Any:
```

Comment:
Is this safe? What happens if we inherit from 2 traits, both defining a `visit`?